graphbolt/src/cuda/index_select_impl.cu

```diff
@@ -48,6 +53,268 @@ std::pair<torch::Tensor, torch::Tensor> Sort(
   return std::make_pair(sorted_array, sorted_idx);
 }
 
+template <typename indptr_t, typename indices_t>
+struct AlignmentFunc {
+  static_assert(GPU_CACHE_LINE_SIZE % sizeof(indices_t) == 0);
+  const indptr_t* in_degree;
+  const int64_t* perm;
+  int64_t num_rows;
+  __host__ __device__ auto operator()(int64_t row) {
+    constexpr int num_elements = GPU_CACHE_LINE_SIZE / sizeof(indices_t);
+    return thrust::make_tuple(
+        in_degree[row],
+        (indptr_t)(in_degree[perm ? perm[row % num_rows] : row] + num_elements - 1));
+  }
+};
+
+template <typename indptr_t, typename indices_t>
+__global__ void _CSRRowWiseOneHopExtractorAlignedKernel(
+    const indptr_t hop_size, const int64_t num_rows,
+    const indptr_t* const indptr, const indptr_t* const subindptr,
+    const indptr_t* const subindptr_aligned, const indices_t* const indices,
+    indices_t* const hop, const int64_t* const perm) {
+  indptr_t tx = static_cast<indptr_t>(blockIdx.x) * blockDim.x + threadIdx.x;
+  const int stride_x = gridDim.x * blockDim.x;
+
+  while (tx < hop_size) {
+    const auto rpos_ = cuda::UpperBound(subindptr_aligned, num_rows, tx) - 1;
+    const auto rpos = perm ? perm[rpos_] : rpos_;
+    const auto out_row = subindptr[rpos];
+    const auto d = subindptr[rpos + 1] - out_row;
+    const int offset =
+        ((size_t)(indices + indptr[rpos] - subindptr_aligned[rpos_]) %
+         GPU_CACHE_LINE_SIZE) /
+        sizeof(indices_t);
+    const auto rofs = tx - subindptr_aligned[rpos_] - offset;
+    if (rofs >= 0 && rofs < d) {
+      const auto in_idx = indptr[rpos] + rofs;
+      assert((size_t)(indices + in_idx - tx) % GPU_CACHE_LINE_SIZE == 0);
+      const auto u = indices[in_idx];
+      hop[out_row + rofs] = u;
+    }
+    tx += stride_x;
+  }
+}
+
+template <typename indptr_t, typename index_t>
+struct DegreeFunc {
+  const index_t* rows;
+  const indptr_t* indptr;
+  indptr_t* in_deg;
+  indptr_t* inrow_indptr;
+  __host__ __device__ auto operator()(int64_t tIdx) {
+    const auto out_row = rows[tIdx];
+    const auto indptr_val = indptr[out_row];
+    const auto degree = indptr[out_row + 1] - indptr_val;
+    in_deg[tIdx] = degree;
+    inrow_indptr[tIdx] = indptr_val;
+  }
+};
+
+struct PairSum {
+  template <typename indptr_t>
+  __host__ __device__ auto operator()(
+      thrust::tuple<indptr_t, indptr_t> a,
+      thrust::tuple<indptr_t, indptr_t> b) {
+    return thrust::make_tuple(
+        thrust::get<0>(a) + thrust::get<0>(b),
+        thrust::get<1>(a) + thrust::get<1>(b));
+  };
+};
+
+std::tuple<torch::Tensor, torch::Tensor> UVAIndexSelectCSCImpl(
```

Comment:
Could you add some comments within the function for readiness. For example, there are several steps in this functions such as index sorting, degree computation, prefix sum, and index copy. Using some lines of comments to separate them would be helpful for code readers.

Reply:
I inserted comments to explain the approach taken, could you check?